WeMo Link add support for tunable white and RGBW lights

jaharkes opened the pull request:
This pretty much rewrites the interface how we talk to zigbee devices through a wemo bridge, so this is really a request for comments.
- Using elementtree to format the setdevicestatus xml request.
- Added support for colortemperature and colorcontrol, see #33  (tested with Osram Lightify RGBW)
- Added per-light/group instances which reduce the number of arguments we need to pass when controlling the lights.
- On the bridge, index devices based on unique device identifier, instead of the user-configured name.

jaharkes committed:
Added temperature and color control.

Lightify RGBW bulbs can be paired with the Link gateway and have
both temperature and RGB color changing functionality.

jaharkes committed:
Create light and group objects

Move linked device operations off of the bridge object.

jaharkes committed:
Work arounds for what seem to be firmware bugs.

When we set brightness to 0 on a bulb that is on it will turn off
and non-zero on an off bulb will turn it back on. But on one of my
bulbs the 'onoff' state doesn't get updated.
Another problem is that a turned off bulb will first turn on to it's
last remembered brightness level and only then fade to the requested
level, so turning on to a dim setting would be a light that switches
on bright and then fades down instead of a slow fade-in.

To work around these,
- If 'sleepfader' is available, use that to turn off the bulb because
  it both dims to 0 over a time interval and updates onoff state.
- When fading from off to on we first force the dim level to the
  lowest non-off state (1), then send a onoff=ON event to update the
  on-off state and finally send the command to fade to the desired
  brightness.

pavoni commented:
The code looks very good. I don't have any Lights - so I can't test myself. 

Your PR says you're looking for comments - but the code looks pretty finished!

Are you happy for me to merge / release a new version - or are you still testing.?

I'll give you access to the repo...

Thanks for the contribution.

jaharkes commented:
Comments because it really changes around how we talk to linked ZigBee devices and drops any backward compatibility with previous versions. OTOH we didn't really have lights working until recently.

Until now I've only been working with this from a interactive Python session directly using the pywemo APIs, I just started modifying the home assistant code to use the new functionality. Cleans up the HA side a bit, and I've already found that the sleepfader control support probably should get folded into the turn_off() function.

On March 13, 2016 3:24:41 PM EDT, Greg Dowling notifications@github.com wrote:

> The code looks very good. I don't have any Lights - so I can't test
> myself. 
> 
> Your PR says you're looking for comments - but the code looks pretty
> finished!
> 
> Are you happy for me to merge / release a new version - or are you
> still testing.?
> 
> I'll give you access to the repo...
> 
> Thanks for the contribution.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/pavoni/pywemo/pull/34#issuecomment-196027313

pavoni commented:
Sounds good - let me know if I can help - and happy to merge when you're ready.

We could build an interim version to make it easier for third parties to test with HA if that helps at some point.

jaharkes commented:
It is a good thing HA has a strong binding to a particular version, so any incompatible release won't accidentally break existing installations.

But yes, being able to test this will be useful.

On March 13, 2016 7:14:30 PM EDT, Greg Dowling notifications@github.com wrote:

> Sounds good - let me know if I can help - and happy to merge when
> you're ready.
> 
> We could build an interim version to make it easier for third parties
> to test with HA if that helps at some point. 
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/pavoni/pywemo/pull/34#issuecomment-196079911

jaharkes committed:
Move sleepfader control into Light.turn_off()

Includes workarounds for 2 more issues.
- sleepfader does not turn off bulb when fadetime is set to 0.
  Forcing minimum fadetime to 1.
- turn_on did not turn bulb on after sleepfade has turned it off.
  Setting brightness with levelcontrol still turns it on, so we
  handle turn_on functionality in the fade up path. i.e. set dim
  level to 0, fire off a turn on signal, and finally fade to the
  requested or last known dim level.

jaharkes committed:
Use GetDeviceStatus call to update status.

GetDeviceStatus gives different results when a light is unreachable, for
instance when it was turned off.
